critical_section.c

```diff
@@ -0,0 +1,24 @@
+#include <critical_section.h>
+#include <interrupts.h>
+#include <common.h>
+#include <libkern.h>
+
+volatile int cs_level = 0;
+
+void cs_enter() {
+  if (during_intr_handler())
+    return;
+  intr_disable();
+  cs_level++;
+}
+
+void cs_leave() {
+  if (during_intr_handler())
+    return;
+
+  assert(cs_level > 0);
+
+  cs_level--;
+  if (cs_level == 0)
+    intr_enable();
```

Comment:
As I have said above, my initial idea was to use these two functions every time interrupts are enabled/disabled (with the first two lines deleted of course). Then this issue wouldn't come up.

Yes, that's true that it would break in that situation, that's a nice catch, thank you!